lib/inspec/run_data.rb

```diff
@@ -0,0 +1,60 @@
+
+module Inspec
+  module HashLikeStruct
+    def keys
+      members
+    end
+
+    def key?(item)
+      members.include?(item)
+    end
+  end
+
+  RunData = Struct.new(
+    :controls,     # Array of Inspec::RunData::Control (flattened)
+    :other_checks,
+    :profiles,     # Array of Inspec::RunData::Profile
+    :platform,     # Inspec::RunData::Platform
+    :statistics,   # Inspec::RunData::Statistics
+    :version       # String
+  ) do
+    include HashLikeStruct
+    def initialize(raw_run_data)
+      self.controls   = raw_run_data[:controls].map { |c| Inspec::RunData::Control.new(c) }
+      self.profiles   = raw_run_data[:profiles].map { |p| Inspec::RunData::Profile.new(p) }
+      self.statistics = Inspec::RunData::Statistics.new(raw_run_data[:statistics])
+      self.platform   = Inspec::RunData::Platform.new(raw_run_data[:platform])
+      self.version    = raw_run_data[:version]
+    end
+  end
+
+  class RunData
+
+    # This is the data layout version of RunData.
+    # We plan to follow a data-oriented version of semver:
+    #  patch: fixing a bug in the provenance or description of a data element, no key changes
+    #  minor: adding new data elements
+    #  major: deleting or renaming data elements
+    # Less than major version 1.0.0, the API is considered unstable.
```

Comment:
This says that < 1.0.0 is unstable, but the PR is starting with version 0.1.0. Do you intend to bump this to 1.0.0 before we merge this PR?

Reply:
@james-stocks and I discussed this a bit, and my thoughts are basically that I feel the API is indeed a bit unstable. This is largely because the source run_data Hash is populated variously depending on the profile being ran, and examined variously depending on the core reporter being used. So, it has been a case of discovery on both ends.

Until the core reporters are adapted to use the RunData object, I'm less confident that there won't be needed changes. That may be non-data changes - like adding more Hash-compatibility features.

That doesn't necessitate blocking this PR on converting the core reporters, though. It does mean that we may want to declare the RunData API as experimental - which it is - and that is a [totally normal thing to do with SemVer](https://semver.org/#spec-item-4).

Alternatively, the RunData API work could be split off, that work incorporated first as an internal refactoring with the core Reporters, then this PR could be revisited with a "mature" API to run against.

Reply:
I'm thinking we can accept this PR as is, with just a little extra explanation for what 0.1.0 means and why we're not 1.0 (yet)

Reply:
Added an explanation.